### PR TITLE
bpo-32002: Refactor C locale coercion tests

### PR DESCRIPTION
**Pull request by ncoghlan**
Exactly which locale requests will end up giving
you the "C" locale is actually platform dependent.

A blank locale and "POSIX" will translate to "C"
on most Linux distros, but may not do so on other
platforms, so this adjusts the way the tests are
structured accordingly.

This may also prove sufficient to fix a current
test failure on Cygwin (hence the issue reference)

https://bugs.python.org/issue32002

### CODE REVIEW

File: Lib/test/test_c_locale_coercion.py
Comment (xdegaye) [inline Lib/test/test_c_locale_coercion.py:18]: I think `""` should be included in this list otherwise the case where all the locale envt variables are set to `""` may not be tested. On Android it is only the tests that have `coerce_c_locale="warn"` that must be skipped when all locale envt variables are set to `""`.

There is currently no test run for when none of these variables is defined and AFAIK it seems correct since it is equivalent to having all locale envt variables set to `""`. But since it is such a common case, testing with all locale envt variables set to `""` should be mandatory for all platforms.

Maybe `EXPECTED_C_LOCALE_EQUIVALENTS` could be replaced with `LOCALES_TO_TEST`.

File: Lib/test/test_c_locale_coercion.py
Comment (xdegaye) [inline Lib/test/test_c_locale_coercion.py]: s/setupModule/setUpModule/

File: Lib/test/test_c_locale_coercion.py
Comment (xdegaye) [inline Lib/test/test_c_locale_coercion.py]: In the comment below:
s/setupModule/setUpModule/

File: Lib/test/test_c_locale_coercion.py
Comment (ncoghlan) [reply Lib/test/test_c_locale_coercion.py:18]: That makes sense. If we go that way, then what I'd actually suggest we do is give the empty locale a dedicated test case, separate from the C locale equivalents.

Comment (xdegaye) [reply Lib/test/test_c_locale_coercion.py:18]: And a dedicated test case for the empty locale would be more explicit.

**Comment by ncoghlan** [comment]
OK, I've factored out the default case to its own subtest. @embray, would you mind trying this patch out on Cygwin and seeing how it breaks?

I'm thinking the right order to do things will be to get https://github.com/python/cpython/pull/4334 merged for Android, rebase this test case change on top of that one, and then figure out a reasonable way to make the expected behaviour in the default locale automatically adjust itself (without making the test case tautological).

File: Lib/test/test_c_locale_coercion.py
Comment (xdegaye) [inline Lib/test/test_c_locale_coercion.py]: s/PYTHON_COERCE_C_LOCALE/PY_COERCE_C_LOCALE/

File: Lib/test/test_c_locale_coercion.py
Comment (xdegaye) [inline Lib/test/test_c_locale_coercion.py]: `base_var_dict` can be used directly without a copy here.

File: Lib/test/test_c_locale_coercion.py
Comment (xdegaye) [inline Lib/test/test_c_locale_coercion.py]: This block can only be run if the following conditional is true, otherwise some test results may not match the expected results:
```
if set(base_var_dict).isdisjoint(extra_vars):
```
For clarity, this statement written above may be moved down, when checking for explicitly configured locales:
```
base_var_dict.update(extra_vars)
```

**Review by xdegaye — CHANGES_REQUESTED** [review]
The tests are successfull on Android when the following changes are made.

File: Lib/test/test_c_locale_coercion.py
Comment (embray) [reply Lib/test/test_c_locale_coercion.py:18]: +1 to a dedicated test case for empty locale.

**Comment by embray** [comment]
I also think that this is one of those cases where *seemingly* tautological test cases may still be worth running, due to non-trivial interactions between the interpreter, the environment, libc, libintl, etc...

**Comment by ncoghlan** [comment]
I'm actually starting to wonder if we should put together a configure check that reports back the result of:

* setting `LC_ALL`, `LC_CTYPE`, and `LANG` to the empty string
* calling `setlocale(LC_CTYPE, "");`
* calling `setlocale(LC_CTYPE, NULL);`

That way we could query `sysconfig` to find out what the build platform did. My main concern with that idea is that it could get complicated when it came to cross-compilation support, so plan B would be to write a short `Programs/_querydefaultlocale.c` test helper that did the same thing.

**Comment by embray** [comment]
This test still fails on Cygwin because it assumes that by default the locale will be "C" and that coercion will happen (which results in the `LC_CTYPE` environment variable getting set).  So these failures look something like:

```
======================================================================
FAIL: test_PYTHONCOERCECLOCALE_not_set (test.test_c_locale_coercion.LocaleCoercionTests) (default_locale=True, PYTHONCOERCECLOCALE=None)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/embray/src/python/cpython/Lib/test/test_c_locale_coercion.py", line 325, in _check_c_locale_coercion
    coercion_expected)
  File "/home/embray/src/python/cpython/Lib/test/test_c_locale_coercion.py", line 221, in _check_child_encoding_details
    self.assertEqual(encoding_details, expected_details)
AssertionError: {'fse[124 chars]slashreplace', 'lang': '', 'lc_ctype': '', 'lc_all': ''} != {'fse[124 chars]slashreplace', 'lang': '', 'lc_ctype': 'c.utf-8', 'lc_all': ''}
  {'fsencoding': 'utf-8',
   'lang': '',
   'lc_all': '',
-  'lc_ctype': '',
+  'lc_ctype': 'c.utf-8',
?               +++++++

   'stderr_info': 'utf-8:backslashreplace',
   'stdin_info': 'utf-8:surrogateescape',
   'stdout_info': 'utf-8:surrogateescape'}
```

In #4361 I just skipped the cases didn't *explicitly* set one of the LC_ vars to `"C"` or `"POSIX"` since otherwise the assumptions in the test of when coercion should occur didn't hold up.

**Comment by embray** [comment]
@ncoghlan I think the configure-time check is a good idea.  It could still be disabled when cross-compiling, in which case a fallback is needed.  But it might be useful to have as a first pass, especially for use by the test suite.

**Comment by embray** [comment]
On problem I have noticed, at least on Cygwin, is that `setlocale()` actually behaves differently depending on whether or not one compiles with libintl 😢.  That's because libintl has its own wrapper around `setlocale()` with different behavior.

**Comment by xdegaye** [comment]
> I'm actually starting to wonder if we should put together a configure check

I think we should limit the use of configure for setting the platform type instead and hard code the specific expectations for a platform in `test_c_locale_coercion.py`. The platform type is needed elsewhere.

For example Android had an entry in the `sysconfig` dictionary at one time (issue 27442) which has been replaced later with the new `sys.getandroidapilevel()` function (issue 28740) to modify `_bootlocale.py` that needs to know very early in the startup sequence whether the platform is Android (issue 28596). Knowing that the platform is Android is also needed in many tests and in `subprocess.py` because its shell is `/system/bin/sh`.

**Comment by xdegaye** [comment]
> This test still fails on Cygwin because it assumes that by default the locale will be "C"

@embray what are the test results if you replace `if test.support.is_android:` with `if True:` in `test_c_locale_coercion.py` with the current implementation of this PR already applied ?

**Comment by embray** [comment]
@xdegaye That fails all the same on Cygwin.  The difference between Cygwin and Android here are different.  On Cygwin the meaning of the `"C"` locale is as in POSIX--it means ASCII encoding.  The only issue on Cygwin is that the system default locale is `"C.UTF-8"` instead of just `"C"` as the tests currently assume.

In other words, `_Py_LegacyLocaleDetected()` is returning `0` on Cygwin in cases where the tests assume it returns `1`.

File: Lib/test/test_c_locale_coercion.py
Comment (ncoghlan) [reply Lib/test/test_c_locale_coercion.py]: The expectation is that the test case that passes in `extra_vars` will also set `expected_warnings` and `coercion_expected` appropriately.

The current use case is the tests that sets `LC_ALL='C'`, and expects that to both disable coercion, and potentially result in a warning (for the `PYTHONCOERCECLOCALE=warn` case).

**Comment by ncoghlan** [comment]
@xdegaye @embray I'm going to merge this once the current CI run finishes (so I can build a patch for https://bugs.python.org/issue30672 on top of it), but I'll leave the tracker issue as "pending" until you've confirmed that this works as intended on Cygwin and Android

**Comment by ncoghlan** [comment]
Although it would help *not* to overwrite the Android fixes when rebasing the patch :)

So change of plan: I'll hold off on merging until you've both given the thumbs up. If it looks good, @xdegaye can merge it, otherwise I'll deal it after I get back from NZ.

**Comment by xdegaye** [comment]
@ncoghlan, @embray

The [test_c_locale_coercion.log](https://github.com/python/cpython/files/1555598/test_c_locale_coercion.log) file is the output of running ``test_c_locale_coercion`` on Android.

``test_c_locale_coercion`` is successful when the [android-bpo-32002.diff.txt](https://github.com/python/cpython/files/1555599/android-bpo-32002.diff.txt) patch is applied on top of Nick last commit (It has a ``.txt`` extension to trick github into accepting it as an attachment).

**Comment by ncoghlan** [comment]
My phone is essentially my only computer until January, but I believe
@xdegaye should have permission to push to my PR branch as a CPython
maintainer.

**Review by xdegaye — APPROVED** [review]

File: Lib/test/test_c_locale_coercion.py
Comment (embray) [inline Lib/test/test_c_locale_coercion.py:338]: I need to study this a bit more carefully, but I believe this condition may apply for Cygwin as well (or anywhere that the default locale has UTF-8 encoding).

**Comment by embray** [comment]
I still get the following failures on Cygwin:

```
======================================================================
FAIL: test_LC_ALL_set_to_C (test.test_c_locale_coercion.LocaleCoercionTests) (default_locale=True, PYTHONCOERCECLOCALE='warn')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/embray/src/python/cpython/Lib/test/test_c_locale_coercion.py", line 344, in _check_c_locale_coercion
    _coercion_expected)
  File "/home/embray/src/python/cpython/Lib/test/test_c_locale_coercion.py", line 235, in _check_child_encoding_details
    self.assertEqual(stderr_lines, expected_warnings)
AssertionError: Lists differ: ['Python runtime initialized with LC_CTYPE[192 chars]ed.'] != []

First list contains 1 additional elements.
First extra element 0:
'Python runtime initialized with LC_CTYPE=C (a locale with default ASCII encoding), which may cause Unicode compatibility problems. Using C.UTF-8, C.utf8, or UTF-8 (if available) as alternative Unicode-compatible locales is recommended.'

+ []
- ['Python runtime initialized with LC_CTYPE=C (a locale with default ASCII '
-  'encoding), which may cause Unicode compatibility problems. Using C.UTF-8, '
-  'C.utf8, or UTF-8 (if available) as alternative Unicode-compatible locales is '
-  'recommended.']

======================================================================
FAIL: test_PYTHONCOERCECLOCALE_set_to_zero (test.test_c_locale_coercion.LocaleCoercionTests) (default_locale=True, PYTHONCOERCECLOCALE='0')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/embray/src/python/cpython/Lib/test/test_c_locale_coercion.py", line 344, in _check_c_locale_coercion
    _coercion_expected)
  File "/home/embray/src/python/cpython/Lib/test/test_c_locale_coercion.py", line 232, in _check_child_encoding_details
    self.assertEqual(encoding_details, expected_details)
AssertionError: {'fsencoding': 'utf-8', 'stdin_info': 'utf-8:surrogateesc[123 chars]: ''} != {'fsencoding': 'ascii', 'stdin_info': 'ascii:surrogateesc[123 chars]: ''}
- {'fsencoding': 'utf-8',
?                 ^^^^^

+ {'fsencoding': 'ascii',
?                 ^^^^^

   'lang': '',
   'lc_all': '',
   'lc_ctype': '',
-  'stderr_info': 'utf-8:backslashreplace',
?                  ^^^^^

+  'stderr_info': 'ascii:backslashreplace',
?                  ^^^^^

-  'stdin_info': 'utf-8:surrogateescape',
?                 ^^^^^

+  'stdin_info': 'ascii:surrogateescape',
?                 ^^^^^

-  'stdout_info': 'utf-8:surrogateescape'}
?                  ^^^^^

+  'stdout_info': 'ascii:surrogateescape'}
?                  ^^^^^
```

But +1 to going ahead and merging, if you want, as this is significantly cleaned up, and I can look into addressing the final Cygwin-related bits on top of this.

**Comment by ncoghlan** [comment]
Merged! Thanks for the reviews and updates, folks :)

**Pull request merged**
